dali/pipeline/util/bounding_box.h

```diff
@@ -20,170 +20,282 @@
+  static BoundingBox FromXywh(float x, float y, float w, float h,
+                              RelBounds limits = {}) {
+    return FromStartAndShape(RelBounds{x, y, w, h}, limits);
+  }
 
-    DALI_ENFORCE(
-      l <= r, "Expected left <= right. Received: " + to_string(l) + " <= " + to_string(r));
-    DALI_ENFORCE(
-      t <= b, "Expected top <= bottom. Received: " + to_string(t) + " <= " + to_string(b));
+  BoundingBox Intersect(const BoundingBox& oth) const {
+    DALI_ENFORCE(ndim() == oth.ndim());
+    RelBounds out;
+    if (!Overlaps(oth)) {
+      return out;
+    }
+    out.reserve(ndim() * 2);
+    for (int i = 0; i < ndim(); i++)
+      out.push_back(std::max(oth.bbox_bounds_[i], bbox_bounds_[i]));
 
-    return {l, t, r, b};
-  }
+    for (int i = 0; i < ndim(); i++)
+      out.push_back(std::min(oth.bbox_bounds_[ndim() + i], bbox_bounds_[ndim() + i]));
 
-  static BoundingBox FromXywh(const float* data, array<float, kSize> bounds = UniformSquare()) {
-    return FromXywh(data[0], data[1], data[2], data[3], bounds);
+    return BoundingBox{out};
   }
 
-  static BoundingBox FromXywh(
-    float x, float y, float w, float h, array<float, kSize> bounds = UniformSquare()) {
-    CheckBounds(x, bounds[0], bounds[2], "x");
-    CheckBounds(y, bounds[0], bounds[2], "y");
-    CheckBounds(x + w, bounds[1], bounds[3], "x + w");
-    CheckBounds(y + h, bounds[1], bounds[3], "y + h");
+  float Volume() const {
+    if (bbox_bounds_.empty())
+      return 0.0f;
 
-    return {x, y, x + w, y + h};
+    float vol = 1.0f;
+    for (int i = 0; i < ndim(); i++) {
+      assert(bbox_bounds_[ndim()+i] >= bbox_bounds_[i]);
+      vol *= bbox_bounds_[ndim()+i] - bbox_bounds_[i];
+    }
+    return vol;
   }
 
-  BoundingBox ClampTo(const BoundingBox& other) const {
-    return {std::max(other.left_, left_), std::max(other.top_, top_),
-            std::min(other.right_, right_), std::min(other.bottom_, bottom_)};
+  float Area() const {
+    DALI_ENFORCE(ndim() == 2,
+      make_string("Requested area on ", ndim(), "-dimensional bounding box. Use Volume() instead"));
+    return Volume();
   }
 
-  float Area() const { return area_; }
+  bool Contains(RelCoords point) const {
+    DALI_ENFORCE(static_cast<int>(point.size()) == ndim(),
+      make_string("Unexpected number of dimensions: ", ndim()));
+    for (int i = 0; i < ndim(); i++) {
+      if (point[i] < bbox_bounds_[i] || point[i] > bbox_bounds_[ndim()+i])
+        return false;
+    }
+    return true;
+  }
 
   bool Contains(float x, float y) const {
-    return x >= left_ && x <= right_ && y >= top_ && y <= bottom_;
+    DALI_ENFORCE(ndim() == 2,
+      make_string("Unexpected number of dimensions: ", ndim()));
+    return Contains(RelCoords{x, y});
   }
 
   bool Overlaps(const BoundingBox& other) const {
-    return left_ < other.right_ && right_ > other.left_ &&
-           top_ < other.bottom_ && bottom_ > other.top_;
+    for (int i = 0; i < ndim(); i++) {
```

Comment:
Can we use `Contains` for that?

Reply:
I don't think it's worth it because I'd have to add corner case handling (that is, the case where both boxes share the edge but they don't overlap)